### PR TITLE
fix(forge): sensitive broadcast logs

### LINKED ISSUE #4730: bug: remove sensitive info from broadcast files
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

forge script

### Operating System

None

### Describe the bug

Broadcast artifacts are intended to be committed so that deployment history can live in a repo, but they currently contain two things that make committing them problematic:
1. The full RPC URL, which exposes API keys, etc.
2. The full, absolute path to the user's project / broadcast file

The RPC URL is required for `--resume` on multi-chain deployments, so we should consider moving it to a different dedicated file that gets written to e.g. the cache folder? (since it's gitignored by default). I'm guessing the full absolute path is there for a similar reason (maybe @joshieDo can confirm) in which case we can use the same solution there

### PR DESCRIPTION
**Pull request by devanoneth**
## Motivation

Remove sensitive values from broadcast logs. Namely `rpc` and `path`. Fixes #4730.

## Solution

As proposed in #4730, saves sensitive values into a file with the same name / folder path but instead of the `broadcast` dir, we use the `cache` dir. It introduces a `ScriptSequenceSensitive` struct which can be used to write sensitive values and later load them back into the `ScriptSequence` struct.

Also fixes and extends the `check_broadcast_log` test so the changes can be tested.

### CODE REVIEW

File: cli/src/cmd/forge/script/sequence.rs [hunk 1]
```diff
@@ -110,6 +169,21 @@ impl ScriptSequence {
...
+            )?;
+            //../run-[timestamp].json
+            serde_json::to_writer_pretty(
+                BufWriter::new(fs::create_file(
+                    sensitive_path.replace("latest.json", &format!("{}.json", self.timestamp)),
+                )?),
+                &script_sequence_sensitive,
+            )?;
+
             shell::println(format!("\nTransactions saved to: {path}\n"))?;
```
Comment (Evalir) [inline cli/src/cmd/forge/script/sequence.rs:187]: maybe we should print an extra message indicating sensitive info has been written to a separate path?

**Review by Evalir — COMMENTED** [review]
Hey hey thanks for this!

logic looks ok, but haven't tested. some nits re docs/comments/naming:

File: cli/src/cmd/forge/script/sequence.rs [hunk 2]
```diff
@@ -45,6 +48,29 @@ pub struct ScriptSequence {
...
 }
 
+#[derive(Deserialize, Serialize, Clone, Default)]
+pub struct TransactionWithMetadataSensitive {
+    pub rpc: Option<String>,
+}
+
+/// Sensitive info from the script sequence which is saved into the cache folder
+#[derive(Deserialize, Serialize, Clone, Default)]
+pub struct ScriptSequenceSensitive {
```
[the remark this answers is not in the document]
Comment (devanoneth) [reply cli/src/cmd/forge/script/sequence.rs:58]: Would `SensitiveScriptSequence` be better? I wanted to keep the name close to the `ScriptSequence` struct as that's where these values come from.

That way you have your `ScriptSequence` which can be committed and `SensitiveScriptSequence` which shouldn't be.

[the remark this answers is not in the document]
Comment (Evalir) [reply cli/src/cmd/forge/script/sequence.rs]: Yep, this sounds good!

[the remark this answers is not in the document]
Comment (devanoneth) [reply cli/src/cmd/forge/script/sequence.rs]: Cool, that's already in my last commit :)

**Comment by Evalir** [comment]
hey hey, will TAL tomorrow and hopefully merge this! looks solid, just haven't fully tested yet

**Comment by devanoneth** [comment]
No prob @Evalir, you're smashing PRs out lately! <3

**Review by Evalir — APPROVED** [review]
this lgtm!
ptal @mattsse

**Review by mattsse — CHANGES_REQUESTED** [review]
cool!

test lgtm!

@Evalir do we need to put this in the changelog?

I believe we need to add this to the default gitignore as well?

**Comment by devanoneth** [comment]
@mattsse thanks for reviewing! I'm not sure why this would need changes in the default `.gitignore`? I just checked and it ignores `cache` and doesn't ignore `broadcast` for any chain without id `31337`:

```
# Compiler files
cache/
out/

# Ignores development broadcast logs
!/broadcast
/broadcast/*/31337/
/broadcast/**/dry-run/

# Docs
docs/

# Dotenv file
.env
```

**Comment by mattsse** [comment]
I see, so the senstive logs are always ignored?

**Comment by devanoneth** [comment]
@mattsse Yeah, the sensitive logs go into directories under the cache directory. See the `get_paths` function in `sequence.rs` for more details.

**Review by mattsse — APPROVED** [review]
gg

**Pull request merged**